### PR TITLE
Slim down schema agreement API

### PR DESCRIPTION
**Pull request by wprzytula**
Schema agreement API is another place in the driver's code that can be polished towards API stability.

## What's done

1. `fetch_schema_version() ` is removed. 
It can be a nasty source of race bugs when schema version is fetched without awaiting for schema agreement. It is especially risky when one awaits schema agreement and subsequently fetches schema version, because the schema version can change in the meantime and schema may no longer be in agreement. To cope with the problem, `await_schema_agreement()` now returns the agreed schema version, and  `fetch_schema_version()` is removed alltogether.
2. `await_schema_agreement()` gets implicit timeout.
    Similarly to
    - what Java driver 3 does in regard to awaiting schema agreement's API,
    - what was recently done in Rust driver in regard to tracing info's API,
    
    the API of awaiting schema agreement is altered so that:
    - `await_timed_schema_agreement()`, which takes explicit timeout, is
      removed,
    - `await_schema_agreement()` is now bound with a timeout that is set
      globally per-`Session` in `SessionConfig`.

    The motivation is that it could lead to application's deadlock when `await_schema_agreement()` was called with no timeout given and the cluster stoped synchronising (so schema agreement would be never reached), It is rarely desirable to block the calling application for arbitrarily long time, especially that some extreme situations are possible, such as a network partition between nodes.
    
3. `SessionBuilder`s API is accomodated to the changes. Docs are updated as well.

### Bonus
- missing docstrings were added for `AddressTranslator`.

Refs: #660 

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~~[ ] I added relevant tests for new features and bug fixes.~~
- [ ] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [x] I have adjusted the documentation in `./docs/source/`.
- ~~[ ] I added appropriate `Fixes:` annotations to PR description.~~

### CODE REVIEW

File: scylla/src/transport/session.rs [hunk 1]
```diff
@@ -1821,77 +1821,33 @@ impl Session {
...
-        Ok(())
     }
 
-    pub async fn await_timed_schema_agreement(
+    pub async fn await_schema_agreement(
         &self,
         timeout_duration: Duration,
-    ) -> Result<bool, QueryError> {
-        timeout(timeout_duration, self.await_schema_agreement())
+    ) -> Result<Option<Uuid>, QueryError> {
```
Comment (avelanarius) [inline scylla/src/transport/session.rs:1836]:
```suggestion
    ) -> Result<Uuid, QueryError> {
```

maybe the timeout should be represented as a timeout `QueryError`?

Comment (wprzytula) [reply scylla/src/transport/session.rs:1836]: Good idea!

Comment (wprzytula) [reply scylla/src/transport/session.rs:1836]: The only problem I see here is that we lose a way to conveniently distinguish cases of timeout (not yet in agreement) from cases of server-side or connection-related errors.

Comment (wprzytula) [reply scylla/src/transport/session.rs]: I've implemented it anyway.

File: docs/source/queries/schema-agreement.md [hunk 2]
```diff
@@ -1,25 +1,14 @@
 # Schema agreement
```
Comment (avelanarius) [inline docs/source/queries/schema-agreement.md:1]: This was a problem before your PR, but this documentation page is missing any mention about automatic schema agreement after schema-changing queries. 

(cc: @muzarski - what we talked about today: the driver does automatic schema agreement after such queries, no need to do it youself)

File: docs/source/queries/schema-agreement.md [hunk 3]
```diff
@@ -87,4 +58,23 @@ if session.check_schema_agreement().await? {
...
+# use std::error::Error;
+# async fn check_only_compiles(session: &Session) -> Result<(), Box<dyn Error>> {
+let session = SessionBuilder::new()
+    .known_node("127.0.0.1:9042")
+    .auto_await_schema_agreement(false)
+    .build()
+    .await?;
+# Ok(())
+# }
+```
```
Comment (avelanarius) [inline docs/source/queries/schema-agreement.md:80]: I think it'd be better if this was at the start of this doc page (as a sort of warning: "You probably don't need to read the remaining portion of the doc - the driver does it for you")

**Review by avelanarius — APPROVED** [review]

File: scylla/src/transport/session.rs [hunk 4]
```diff
@@ -83,6 +83,24 @@ use crate::authentication::AuthenticatorProvider;
...
+ * directly by the driver or should not be the preferred address to use to reach the node (for
+ * instance, the `rpc_address` set on ScyllaDB nodes might be a private IP, but some clients
+ * may have to use a public IP, or pass by a router, e.g. through NAT, to reach that node).
+ * This interface allows to deal with such cases, by allowing to translate an address as sent
+ * by a ScyllaDB node to another address to be used by the driver for connection.
+ *
+ * Please note that the "known nodes" addresses provided while creating the [`Session`]
+ * instance are not translated, only IP address retrieved from or sent by Cassandra nodes
+ * to the driver are.
+ */
```
Comment (piodul) [inline scylla/src/transport/session.rs:103]: How is this related to slimming down schema agreement API? This should be sent separately.

**Review by piodul — APPROVED** [review]

**Comment by piodul** [comment]
Requires a rebase and dropping one of the commits, but otherwise LGTM.

File: scylla/src/transport/session.rs (hunk 4 above)
Comment (wprzytula) [reply scylla/src/transport/session.rs:103]: This is a bonus, as described in the PR's description. Opening PRs have nontrivial overhead, so I prefer to add such tiny things to existing ones.

Comment (piodul) [reply scylla/src/transport/session.rs:103]: > Opening PRs has nontrivial overhead

Is it really that much of an overhead to send a single commit PR? You send it, then somebody reviews it quickly (as it usually is with such simple improvements) and later we can all forget about it before the rest is reviewed.

> I prefer to add such tiny things to existing ones.

No, please, for the sake of order let's not mix up things. Each PR should be concentrated on one topic.

**Comment by wprzytula** [comment]
Rebased and got rid of irrelevant commit.

**Comment by piodul** [comment]
Thanks. I'll merge after the CI checks finish.
